### DISCUSSION TITLE
Git submodules and imports

### DISCUSSION THREAD

**Discussion by ampx-mg** [apache/airflow#35673] [Q&A] [answered]
Hi,

I ran into this discussion recently https://github.com/apache/airflow/discussions/19381#discussioncomment-2509788, because I am solving a similar problem.  
We have an umbrella repo - `airflow_jobs` which we add all git submodules to, say `airflow_project_1`, `airflow_project_2`, etc. My dags_folder is setup to `/opt/airflow/dags/repo/airflow_jobs/` and everything works great.  
Now, I would like to have `utils` folder in `airflow_project_1` for every DAG to be importable and usable. However I am struggling to find the proper way how to do it.
- relative imports are discouraged by airflow docs
- having `airflow_project_1/dags` and `airflow_project_1/utils` and using `from utils import ...` works in the project repo, but the path is obviously different in the umbrella repo. I would like to have `utils` folder in every single project
- turning the git submodule part with utils into package could work, but I dont want to install any extra packages on webserver/scheduler and doing this for every single project doesn't scale well

It is important for me to preserve seamless development environment for our data scientists, meaning they should be able to run unittests locally without additional issues to worry about. 

Has anybody dealt with this problem in an elegant way they would be willing to share?

**Comment by potiuk** [comment, marked as answer]
The problem you have that you cannot use "from utils" in two separate projects - because you would not know which "utils" you want to import. Obviously the relative imports would be best in this case - but I am afraid the way we are importing it will make them now work (you can try - but I think it will not work.

I think the best idea is to have for your data scientist to:

1) let all your data scientist check out the project repos somewhere which they could name (say) "workspace" (this will be equivalent of the "AIRFLOW_DAG" folder
2) when they check-out - they shoul  use `airflow_project_1" as checked out repository name (which it should be pretty much default if your project in git is named like that)
3) make sure you have `__init__.py" in both root of your "airflow_project_1" and "utils" subfolder (which you should have anyway) 
4) make sure (and that's something to figure out for any way your "managed" dev environment is) that the `workspace` is added to the PYTHONPATH of their - could be via env variable, configuration of the IDE for example. It could also be .pth file with link to it in the virtualenv they use, there are few other options I think that you can automate and it should be linked with the way how you let them manage their virtualenvs.

This way the "workspace" structure will look like this:

```
workspace <--- HAVE THIS ADDED TO PYTHONPATH
        |-airflow_project_1
        |                  \--utils
        |-airflow_project_2
        |                  \--utils
 ```  

This way your data scientists shoudl use `from airflow_project_1.utils import` - same as in `airflow` when it runs.

This also really nice if you want to share some "common utitilies" between different projects because you can also do that:

```
workspace <--- HAVE THIS ADDED TO PYTHONPATH
        |-airflow_project_1
        |                  \--utils
        |-airflow_project_2
        |                  \--utils
        |-common_library_for_all_project
        |                               \--utils
 ```  

And the only requirement is that data scientists of your must check it out in their workspace (usually separate repo) and use it via `from common_library_for_all_project.utils import ....`

**Reply by ampx-mg** [reply]
Thanks for the detailed answer.

Tried it out and finally managed to get it working both locally and then on our CICD pipelines.

**Reply by potiuk** [reply]
Glad I could help :)
